store/snapshots/store_test.go

```diff
@@ -318,19 +320,43 @@ func TestStore_Save(t *testing.T) {
 
 	// Saving a snapshot should error if a snapshot is already in progress for the same height,
 	// regardless of format. However, a different height should succeed.
-	ch = make(chan io.ReadCloser)
-	mtx := sync.Mutex{}
-	mtx.Lock()
-	go func() {
-		mtx.Unlock()
-		_, err := store.Save(7, 1, ch)
-		require.NoError(t, err)
-	}()
-	mtx.Lock()
-	defer mtx.Unlock()
-	_, err = store.Save(7, 2, makeChunks(nil))
-	require.Error(t, err)
-	_, err = store.Save(8, 1, makeChunks(nil))
-	require.NoError(t, err)
-	close(ch)
+	var (
+		wgStart, wgDone sync.WaitGroup
+		errCount        atomic.Uint32
+	)
+	const n = 3
+	wgStart.Add(n + 1)
+	wgDone.Add(n + 1)
+	for i := 0; i <= n; i++ {
+		ch = make(chan io.ReadCloser, 1)
+		ch <- &ReadCloserMock{} // does not block on a buffered channel
+		close(ch)
+		go func(i int) {
+			wgStart.Done()
+			wgStart.Wait() // wait for all routines started
+			var err error
+			if i < n {
+				_, err = store.Save(7, 1, ch)
+			} else {
+				_, err = store.Save(8, 1, makeChunks(nil))
```

Comment:
good intend but this condition on `n` does not make it very readable, IMHO.
Either you can cover the second test case in separate code block or you refactor towards the height as argument. For example:
```go
	var (
		wgStart, wgDone sync.WaitGroup
		mu              sync.Mutex
		gotErrHeights   []uint64
	)
	srcHeights := []uint64{7, 7, 7, 8, 9}
	wgStart.Add(len(srcHeights))
	wgDone.Add(len(srcHeights))
	for _, h := range srcHeights {
		ch = make(chan io.ReadCloser, 1)
		ch <- &ReadCloserMock{} // does not block on a buffered channel
		close(ch)
		go func(height uint64) {
			wgStart.Done()
			wgStart.Wait() // wait for all routines started
			if _, err = store.Save(height, 1, ch); err != nil {
				mu.Lock()
				gotErrHeights = append(gotErrHeights, height)
				mu.Unlock()
			}
			wgDone.Done()
		}(h)
	}
	wgDone.Wait() // wait for all routines completed
	assert.Equal(t, []uint64{7, 7}, gotErrHeights)
```
This approach has the benefit that you can identify the failing height rather than relying on the count. WDYT?

Reply:
fixed